crates/mpz-garble/src/protocol/semihonest/generator.rs

```diff
@@ -0,0 +1,319 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use tokio::sync::Mutex;
+use utils::filter_drain::FilterDrain;
+
+use mpz_common::{
+    scoped_futures::{ScopedBoxFuture, ScopedFutureExt},
+    Context, Flush,
+};
+use mpz_core::{bitvec::BitVec, Block};
+use mpz_garble_core::GeneratorOutput;
+use mpz_memory_core::{binary::Binary, correlated::Delta, DecodeFuture, Memory, Slice, View};
+use mpz_ot::cot::COTSender;
+use mpz_vm_core::{Call, Execute, Vm};
+
+use crate::store::{GeneratorStore, GeneratorStoreError};
+
+type Result<T, E = GeneratorError> = core::result::Result<T, E>;
+type Error = GeneratorError;
+
+/// Semi-honest generator.
+#[derive(Debug)]
+pub struct Generator<COT> {
+    store: Arc<Mutex<GeneratorStore<COT>>>,
+    call_stack: Vec<(Call, Slice)>,
+    preprocessed: Vec<(Vec<Slice>, Slice)>,
+}
+
+impl<COT> Generator<COT> {
+    /// Creates a new generator.
+    pub fn new(cot: COT, seed: [u8; 16], delta: Delta) -> Self {
+        Self {
+            store: Arc::new(Mutex::new(GeneratorStore::new(seed, delta, cot))),
+            call_stack: Vec::new(),
+            preprocessed: Vec::new(),
+        }
+    }
+
+    fn take_preprocess_calls(&mut self) -> Vec<(Call, Slice)> {
+        let store = self.store.try_lock().unwrap();
+        self.call_stack
+            .filter_drain(|(call, _)| call.inputs().iter().all(|slice| store.is_set_keys(*slice)))
+            .collect()
+    }
```

Comment:
On the evaluator side you filter for calls which are independent. Why don't you have to do that here?